manual.adoc

```diff
@@ -211,5 +211,112 @@ In the example above, we've created two registries:
 
 Both registry have a `[shout]` block extension registered with a specific implementation.
 
-The first block extension is registered in the `registryA` and will transform the text to uppercase. 
+The first block extension is registered in the `registryA` and will transform the text to uppercase.
 The other one is registered in the `registryB` and will replace `.` by `{nbsp}!!!`.
+
+=== How to compile a Ruby extension to JavaScript (Node.js)
+
+Let's say you want to use the following Asciidoctor extension (written in Ruby) in your JavaScript based application.
+
+.lib/tel-inline-macro.rb
+```ruby
+require 'asciidoctor/extensions' unless RUBY_ENGINE == 'opal' # <1>
+
+include ::Asciidoctor
+
+# An inline macro that turns URIs with the tel: URI scheme into links.
+#
+# Usage
+#
+#   tel:1-800-555-1212[]
+#
+Extensions.register do
+  inline_macro do
+    named :tel
+    parse_content_as :text
+    process do |parent, target, attrs|
+      if (text = attrs['text']).empty?
+        text = target
+      end
+      target = %(tel:#{target})
+      (create_anchor parent, text, type: :link, target: target).render
+    end
+  end
+end
+```
+<1> Conditional `require` because the compiler can not resolve this dependency but Asciidoctor.js will provide it at runtime
+
+To compile this extension, you will need to install the `opal-compiler` package:
+
+ npm install opal-compiler
+
+Once this package is installed, you can write the following Node script to compile the extension:
+
+.compile.js
+```js
+const fs = require('fs');
+const Builder = require('opal-compiler').Builder;
+// Opal object will be available on the global scope
+
+const builder = Builder.$new();
+builder.$append_paths('lib'); // <1>
+const result = builder.$build('tel-inline-macro.rb'); // <2>
+fs.writeFileSync('tel-inline-macro.js', result.$to_s(), 'utf8'); // <3>
+```
+<1> The extension file is in the `lib` directory
+<2> The extension file is named `tel-inline-macro.rb`
+<3> A file named `tel-inline-macro.js` will be written
+
+Then run it:
+
+ node compile.js
+
+The above command will create a file named `tel-inline-macro.js` containing the extension code in JavaScript.
+
+=== How to use an Asciidoctor extension
+
+In the previous section, we showed how to compile a Ruby extension to JavaScript.
+In this section we will use this extension.
+
+Let's create a simple document:
+
+.document.adoc
+```adoc
+= Contacts
+
+== Professional
+
+Guillaume Grossetie tel:1-800-555-1212[]
+```
+
+Next we create a Node script to convert this document:
+
+.convert.js
+[source,javascript]
+----
+const fs = require('fs');
+const asciidoctor = require('asciidoctor.js')();
+require('./tel-inline-macro.js'); // <1>
+
+const html = asciidoctor.convert(fs.readFileSync('document.adoc', 'utf-8')); // <2>
```

Comment:
Perhaps use:

```
const html = asciidoctor.convertFile('document.adoc', { to_file: false })
```

Reply:
Good idea, `require('fs')` is now longer needed.